sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala

```diff
@@ -35,14 +35,25 @@ case class CoalesceShufflePartitions(session: SparkSession) extends CustomShuffl
     if (!conf.coalesceShufflePartitionsEnabled) {
       return plan
     }
-    if (!plan.collectLeaves().forall(_.isInstanceOf[QueryStageExec])
-        || plan.find(_.isInstanceOf[CustomShuffleReaderExec]).isDefined) {
-      // If not all leaf nodes are query stages, it's not safe to reduce the number of
-      // shuffle partitions, because we may break the assumption that all children of a spark plan
-      // have same number of output partitions.
-      return plan
+
+    if (canCoalescePartitions(plan)) {
+      coalescePartitions(plan)
+    } else {
+      plan.transformUp {
+        case u: UnionExec =>
+          u.withNewChildren(u.children.map { child =>
+            if (canCoalescePartitions(child) &&
+              child.find(_.isInstanceOf[UnionExec]).isEmpty) {
```

Comment:
We still need this check `child.find(_.isInstanceOf[UnionExec]).isEmpty` ? It seems `canCoalescePartitions(child)` always return false if `child.find(_.isInstanceOf[UnionExec]).isEmpty` is false?

Reply:
We should add the coalesce if it's children don't have `Union` to avoid adding duplicate `CustomShufflerReader`.

Without `CombineUnions`, the plan can be
````
Union
  Union
     HashAggregate
      ShuffleQueryStage
     FileScan
  FileScan
````